src/core/mutatus.scala

```diff
@@ -319,8 +313,88 @@ object Dao {
       implicit metadata: TypeMetadata[T],
       decoder: Decoder[T],
       namespace: Namespace,
-      service: Service
+      ctx: Context
   ): Dao[T] = Dao(metadata.typeName)
+
+  /**
+    * Executes operations from tx body using Datastore Transaction API.
+    * If all operations would pass transaction would be automaticlly commited, in otherwise it would be rolledback
+    *
+    * @param tx Body of transaction to be performed. Input parameter must be used as implicit value in order to work.
+    *
+    * Example of usage
+    * Dao.transaction{ implicit tx =>
+    *   for{
+    *     foo <- Dao.all.filter(_.bar == "bar").run()
+    *     bar  <- foo.map(_.copy(bar = "foo")).saveAll
+    *   } yield bar
+    * }
+    */
+  def transaction[T](
+      tx: Context.Transaction => Result[T]
+  )(implicit svc: Service) = {
+    val ctx = Context.Transaction(svc)
+    tx(ctx)
+      .flatMap { result =>
+        Result(ctx.tx.commit())
+          .map(_ => result)
+          .extenuate {
+            case ex: DatastoreException => DatabaseException(ex)
+          }
+      }
+      .extenuate {
+        case ex: DatastoreException =>
+          if (ctx.tx.isActive) {
+            ctx.tx.rollback()
+          }
+          DatabaseException(ex)
+      }
+  }
+ 
+
+  def batchImpl[T: c.WeakTypeTag](
+      c: blackbox.Context
+  )(batch: c.Tree)(svc: c.Tree): c.Tree = {
+    import c.universe._
+    val q"($arg) => $fnBody" = batch
+    //Using external context is a proof that provided Context.Batch was not sufficient to perform operation, though external (default) context was used
+    //Such construct could be considered as dirty hack, but I've not found  better way to found usage of default context
+    val usesExternalContext = fnBody.exists(showCode(_).contains("Context.default"))
```

Comment:
This one is a bit dirty hack, but I've not found better solution to check at compile time if prohibited read operations were used.